openedx/openedx-platform
# Several enhancements and refactors to CourseGraph

## Description

In order to learn [Tutor](https://tutor.overhang.io), I began working on a CourseGraph plugin for Tutor. The work-in-progress plugin is [here](https://github.com/kdmccormick/tutor-contrib-coursegraph). In doing so, a couple CourseGraph pitfalls became apparent:
* The management command for refreshing CourseGraph from CMS (`./manage.py cms dump_to_neo4j`) was very clunky: it had several arguments that needed to be specified every time the command was called (instead of falling back to reasonable defaults). Furthermore, it did not provide documentation for those arguments to command-line users; one would have to go diving in the code to find details on each argument.
* The management command was advertised as available from both LMS and CMS context, but only worked reliably across environments in a CMS context.
* Needing to run a management command to push courses to CourseGraph at all seemed like an anti-pattern. It would be much easier for operators if CourseGraph could be refreshed from Django admin, or better yet, completely automatically whenever a course is published.

This PR fixes all three of those issues across six commits, which I recommend reviewing individually. Each commit includes more technical details in its body.

This PR depends on https://github.com/openedx/devstack/pull/899.

## Supporting information

New `dump_to_neo4j` command-line helptext:
```
app@2a67edc990cc:~/edx-platform$ ./manage.py cms dump_to_neo4j --help

(.... CMS startup noise ...)

usage: manage.py dump_to_neo4j [-h] [--host HOST] [--port PORT] [--secure] [--user USER] [--password PASSWORD] [--courses [KEY [KEY ...]]] [--skip [KEY [KEY ...]]] [--override] [--version] [-v {0,1,2,3}]
                               [--settings SETTINGS] [--pythonpath PYTHONPATH] [--traceback] [--no-color] [--force-color] [--skip-checks]

Dump recently-published course(s) over to a CourseGraph (Neo4j) instance.

optional arguments:
  -h, --help            show this help message and exit
  --host HOST           the hostname of the Neo4j server
  --port PORT           the port on the Neo4j server that accepts Bolt requests
  --secure              connect to server over Bolt/TLS instead of plain unencrypted Bolt
  --user USER           the username of the Neo4j user
  --password PASSWORD   the password of the Neo4j user
  --courses [KEY [KEY ...]]
                        keys of courses to serialize; if omitted all courses in system are serialized
  --skip [KEY [KEY ...]]
                        keys of courses to NOT to serialize
  --override            dump all courses regardless of when they were last published
  --version             show program's version number and exit
  -v {0,1,2,3}, --verbosity {0,1,2,3}
                        Verbosity level; 0=minimal output, 1=normal output, 2=verbose output, 3=very verbose output
  --settings SETTINGS   The Python path to a settings module, e.g. "myproject.settings.main". If this isn't provided, the DJANGO_SETTINGS_MODULE environment variable will be used.
  --pythonpath PYTHONPATH
                        A directory to add to the Python path, e.g. "/home/djangoprojects/myproject".
  --traceback           Raise on CommandError exceptions
  --no-color            Don't colorize the command output.
  --force-color         Force colorization of the command output.
  --skip-checks         Skip system checks.
app@2a67edc990cc:~/edx-platform$ 
```

The new admin interface, showing a CourseOverviews-backed table, from which sets of courses can be dumped to CourseGraph:
![image](https://user-images.githubusercontent.com/3628148/153106921-0e8c404b-df88-4c15-afbe-26627873d43e.png)

Messaging when successful:
![image](https://user-images.githubusercontent.com/3628148/153107016-fc6354d8-1c61-4728-b0a4-59150a3bf7b2.png)

Messaging when there's nothing to do:
![image](https://user-images.githubusercontent.com/3628148/153107219-070cee32-13a2-42eb-b718-381dbe51cde3.png)

Messaging when there was an error (eg, bad Neo4j connection parameters):
![image](https://user-images.githubusercontent.com/3628148/153107269-fc86d57c-ed0c-4145-989c-9c9917a309ad.png)


## Testing instructions

Here are some test suggestions, although I encourage you to mess around with the new features.

### Setup

Check out this PR's edx-platform branch.
Check out this devstack branch: https://github.com/openedx/devstack/pull/899

From devstack, run `make dev.provision.coursegraph`.
Visit http://localhost:7474, and log into Neo4j with username `neo4j` and password `edx`.

In Studio, find a course you can make some arbitrary edits to.

### Test management command 

* Run `./manage.py cms dump_to_neo4j`. 
* From Neo4j, run a query. For example, this one just counts all blocks in all courses: `MATCH (b) RETURN count(b)`.
* Add a block to your course and publish.
* Run your query again. The change shouldn't be manifested in CourseGraph yet.
* Run `./manage.py cms dump_to_neo4j` again.
* Run your query again. The change should be manifested in CourseGraph.

### Test admin action

* Go to Studio admin and find "Course graph course dumps".
* Select your course and dump it (respecting cache). You should be notified that your course was skipped.
* Add a block to your course and publish.
* Dump your course again. You should be notified that a dump was enqueued for your course.
* Dump your course again (overriding cache). You should be notified that a dump was enqueued for your course.
* Bring down coursegraph: `make dev.down.coursegraph`.
* Dump your course again. You should be notified that a connection error occurred.
* Bring coursegraph back up: `make dev.up.coursegraph`.

### Test auto push

* Edit the file `cms/envs/private.py`, creating it if it doesn't exist.
* Add `COURSEGRAPH_DUMP_COURSE_ON_PUBLISH = True`. Reboot CMS: `make studio-restart-devserver`.
* Add a block to your course and publish.
* Run your CourseGraph query again. The change should be manifested in CourseGraph without having to use the management command or the admin interface!


## Other information

### Blocker: devstack PR

This small devstack PR needs to be reviewed & merged first: https://github.com/openedx/devstack/pull/899

### edX operational simplifications

This change would allow 2U-OCM/edX to make some operational simplifications if they wanted to (although their current setup should keep working fine as-is):
* By configuring `COURSEGRAPH_CONNECTION` in prod-edx-edxapp-cms, the new admin interface could be used to backfill courses into CourseGraph whenever necessary.
* By further enabling `COURSEGRAPH_DUMP_ON_COURSE_PUBLISH` in prod-edx-edxapp-cms, courses would be automatically & continuously pushed to CourseGraph upon publish.

Together, these changes would allow edX to decommission their CourseGraph Jenkins jobs.

### The README

The [CourseGraph readme](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/coursegraph/README.rst) will still be accurate after this change, but it'll be missing the new capabilities that this PR adds. I plan on overhauling it all in one go once the CourseGraph Tutor plugin is ready to go.
